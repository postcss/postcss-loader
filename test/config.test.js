import path from 'path';

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from './helpers';

const testDirectory = path.resolve(__dirname, './fixtures', 'config-autoload');

describe('"config" option', () => {
  it('should work without the specified value', async () => {
    const compiler = getCompiler('./config-scope/css/index.js');
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work without the specified value in the "postcssOptions" option', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {},
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "false" value', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: false,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "true" value', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: true,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "String" value (absolute path)', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          './fixtures/config-scope/css/custom.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work "String" value (relative path)', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: 'test/fixtures/config-scope/css/custom.config.js',
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "String" value (with path to the directory with the configuration)', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: 'test/fixtures/config-scope',
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "package.json" configuration', async () => {
    const compiler = getCompiler('./config-autoload/pkg/index.js', {
      postcssOptions: {
        config: path.resolve(testDirectory, 'pkg'),
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('index.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with the "postcssOptions" option', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: true,
        plugins: ['postcss-rtl'],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "String" value and respect all options', async () => {
    const compiler = getCompiler('./sss/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          './fixtures/config-scope/all-options/postcss.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.sss', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should throw an error on the unresolved config', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          './fixtures/config-scope/css/unresolve.js'
        ),
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });

  it('should throw an error on the invalid config', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          './fixtures/config-scope/css/invalid.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });

  it('should work and resolve "from" and "to" options', async () => {
    const compiler = getCompiler('./config-scope/css/index.js', {
      postcssOptions: {
        config: path.resolve(
          __dirname,
          './fixtures/config-scope/from-to/postcss.config.js'
        ),
      },
    });
    const stats = await compile(compiler);

    const { css, sourceMap } = getCodeFromBundle('style.css', stats);

    sourceMap.sourceRoot = '';
    sourceMap.sources = sourceMap.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(false);
      expect(source).toBe(path.normalize(source));

      return source.replace(/\\/g, '/');
    });

    expect(css).toMatchSnapshot('css');
    expect(sourceMap).toMatchSnapshot('source map');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });
});
