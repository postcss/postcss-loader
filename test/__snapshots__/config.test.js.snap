// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"config" option should throw an error on the invalid config: errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: invalid postcss config",
]
`;

exports[`"config" option should throw an error on the invalid config: warnings 1`] = `Array []`;

exports[`"config" option should throw an error on the unresolved config: errors 1`] = `
Array [
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: No PostCSS Config found in: /test/fixtures/config-scope/css/unresolve.js",
]
`;

exports[`"config" option should throw an error on the unresolved config: warnings 1`] = `Array []`;

exports[`"config" option should work "String" value (relative path): css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work "String" value (relative path): errors 1`] = `Array []`;

exports[`"config" option should work "String" value (relative path): warnings 1`] = `Array []`;

exports[`"config" option should work resolve "from" and "to" options: css 1`] = `
"a { color: black }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work resolve "from" and "to" options: errors 1`] = `Array []`;

exports[`"config" option should work resolve "from" and "to" options: source map 1`] = `
Object {
  "file": "style.css",
  "mappings": "AAAA,IAAI,aAAa;;AAEjB;EACE,YAAY;AACd",
  "names": Array [],
  "sourceRoot": "",
  "sources": Array [
    "style.css",
  ],
  "sourcesContent": Array [
    "a { color: black }

.foo {
  float: right;
}
",
  ],
  "version": 3,
}
`;

exports[`"config" option should work resolve "from" and "to" options: warnings 1`] = `Array []`;

exports[`"config" option should work with "String" value (absolute path): css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work with "String" value (absolute path): errors 1`] = `Array []`;

exports[`"config" option should work with "String" value (absolute path): warnings 1`] = `Array []`;

exports[`"config" option should work with "String" value (with path to the directory with the configuration): css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work with "String" value (with path to the directory with the configuration): errors 1`] = `Array []`;

exports[`"config" option should work with "String" value (with path to the directory with the configuration): warnings 1`] = `Array []`;

exports[`"config" option should work with "String" value and respect all options: css 1`] = `
"a {
  color: black
}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlLnNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQTtFQUNFO0FBREYiLCJmaWxlIjoic3R5bGUuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiYVxuICBjb2xvcjogYmxhY2tcbiJdfQ== */"
`;

exports[`"config" option should work with "String" value and respect all options: errors 1`] = `Array []`;

exports[`"config" option should work with "String" value and respect all options: warnings 1`] = `Array []`;

exports[`"config" option should work with "false" value: css 1`] = `
"a { color: black }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work with "false" value: errors 1`] = `Array []`;

exports[`"config" option should work with "false" value: warnings 1`] = `Array []`;

exports[`"config" option should work with "package.json" configuration: css 1`] = `
".import {
  color: red;
}

.test {
  color: blue;
}
"
`;

exports[`"config" option should work with "package.json" configuration: errors 1`] = `Array []`;

exports[`"config" option should work with "package.json" configuration: warnings 1`] = `Array []`;

exports[`"config" option should work with "true" value: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work with "true" value: errors 1`] = `Array []`;

exports[`"config" option should work with "true" value: warnings 1`] = `Array []`;

exports[`"config" option should work with the "postcssOptions" option: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

[dir=ltr] .foo { float: right;
}

[dir=rtl] .foo {
  float: left;
}
"
`;

exports[`"config" option should work with the "postcssOptions" option: errors 1`] = `Array []`;

exports[`"config" option should work with the "postcssOptions" option: warnings 1`] = `Array []`;

exports[`"config" option should work without the specified value in the "postcssOptions" option: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work without the specified value in the "postcssOptions" option: errors 1`] = `Array []`;

exports[`"config" option should work without the specified value in the "postcssOptions" option: warnings 1`] = `Array []`;

exports[`"config" option should work without the specified value: css 1`] = `
"a { color: rgba(0, 0, 0, 1.0) }

.foo {
  float: right;
}
"
`;

exports[`"config" option should work without the specified value: errors 1`] = `Array []`;

exports[`"config" option should work without the specified value: warnings 1`] = `Array []`;
